bob-common/src/mapper.rs

```diff
@@ -121,28 +119,98 @@ impl Virtual {
+        if res.index == None && res.avail > 0 {
+            let mut curr_offset = offset % res.avail;
+            while res.index == None && res.avail > 0 {
+                res = Self::try_find_support_node_offset_at_one_pass(
+                    nodes,
+                    target_nodes,
+                    curr_offset,
+                );
+                curr_offset = if curr_offset > res.avail {
+                    curr_offset - res.avail
+                } else {
+                    0
+                }
+            }
+        }
+        match res.index {
+            Some(index) => index,
+            None => offset % nodes.len(),
+        }
+    }
+
     pub fn get_support_nodes(&self, key: BobKey, count: usize) -> Vec<&Node> {
         debug_assert!(count <= self.nodes.len());
         if count == 0 {
             return vec![];
         }
         trace!("get target nodes for given key");
         let target_nodes = self.get_target_nodes_for_key(key);
+        debug_assert!(target_nodes.iter().map(|n| n.index()).collect::<HashSet<NodeId>>().len() == target_nodes.len());
+        if target_nodes.len() >= self.nodes.len() {
+            return vec![];
+        }
         trace!("extract indexes of target nodes");
         trace!("nodes available: {}", self.nodes.len());
-        let offset = self.support_nodes_offset.fetch_add(1, Ordering::Relaxed);
-        let mut support_nodes = Vec::with_capacity(count);
-        for (id, node) in self.nodes.iter().skip(offset % self.nodes.len()) {
-            if target_nodes.iter().all(|i| i.index() != *id) {
+        let mut support_nodes: Vec<&Node> = Vec::with_capacity(count);
+
+        // This offset search preserves the uniform distribution of aliens
+        let starting_index = Self::find_support_node_offset(
+            &self.nodes,
+            target_nodes,
+            self.support_nodes_offset.fetch_add(1, Ordering::Relaxed)
+                % (self.nodes.len() - target_nodes.len()),
+        );
+
+        let len = self.nodes.len();
+        for i in 0..len {
+            let node = &self.nodes[(i + starting_index) % len];
+            if node.connection_available() && target_nodes.iter().all(|n| n.index() != node.index())
+            {
                 support_nodes.push(node);
                 if support_nodes.len() >= count {
                     break;
                 }
             }
         }
-        if support_nodes.len() < count && offset % self.nodes.len() > 0 {
-            for (id, node) in self.nodes.iter().take(offset % self.nodes.len()) {
-                if target_nodes.iter().all(|i| i.index() != *id) {
+        if support_nodes.len() < count
+            && support_nodes.len() + target_nodes.len() < self.nodes.len()
+        {
+            for i in 0..len {
+                let node = &self.nodes[(i + starting_index) % len];
+                // Ignore connection status to fill support_nodes
```

Comment:
Please, explain why the connection status is ignored (this is because we are trying to fill the support nodes vector up to the requested `count` of nodes)